### DISCUSSION TITLE
Too slow speed for lammps gpu

### DISCUSSION THREAD

**Discussion by Juo-kim** [ACEsuit/mace#562] [Q&A]
Hello,

I am running the mace LAMMPS gpu , but the computation speed seems unusually low. I am using an RTX 3090 GPU, and I will attach the log file and the file I used for compilation.

The structure I used contains only 12 atoms, so I am wondering if this is the expected performance.

For the MACE model, I downloaded and used the one named "2024-01-07-mace-128-L2_epoch-199.model-lammps" from MACE-MP.

And this is my input file
####input#####
units         metal
boundary      p p p
atom_style    atomic
atom_modify map yes
newton on

read_data     ./dat.lammps

pair_style mace no_domain_decomposition
#pair_coeff      * * ./2023-12-03-mace-mp.model-lammps.pt  Zr O
pair_coeff      * * ./2024-01-07-mace-128-L2_epoch-199.model-lammps.pt  Zr O 
dump          myDump all custom 10 xyz.lammpstrj id element x y z
dump_modify   myDump sort id element Zr O

thermo_style  custom step time cpu pe ke etotal temp press vol density
thermo        10

velocity      all create 300.0 12345
fix           myEnse all npt temp 300.0 300.0 0.1 aniso 1.0 1.0 1.0
timestep      0.002
run           10
################
[lammps.log](https://github.com/user-attachments/files/16694186/lammps.log)

[lammps-mace compile.txt](https://github.com/user-attachments/files/16694166/lammps-mace.compile.txt)

**Comment by gabor1** [comment]
You need to run a few hundred maybe a thousand steps to get a good sense of speed, because the neighbour list calculation is not done on every step.

**Reply by gabor1** [reply]
Others are better placed to comment on the actual speed, but based on common benchmarks I would expect the RTX 3090 to be able half as fast as the A100. You are using the large model (L2), so if I translate that to our runs, getting about half a ns/day on the A100 does not strike me as way off from what we expect. 

The other comment I have is that because your system is very small and periodic, you have a huge number of ghost atoms, and that dominates the calculation. If you made the system bigger (I don't know if this makes sense, I don't know what your application is), I would expect that initially the time taken would go up less than linearly.

**Reply by wcwitt** [reply]
Would you please try again, running 1000 steps, and send us the log? It is normal for the first several steps to be extremely slow while Torch is adjusting the sizes of its internal arrays.

> you have a huge number of ghost atoms, and that dominates the calculation

This is indeed a factor, but in `no_domain_decomposition` mode (which they are using) the situation is much improved because the ghost atoms are mapped back to local atoms.

**Reply by Juo-kim** [reply]
@gabor1 
Thank you for your response.

Ah, this was just a test, and in reality, I have performed calculations with more steps. For now, I am providing the calculation results for 1000 steps. In actual calculations, I work with structures that contain between 100 and 200 atoms. However, the calculations seem to be very slow even for those structures, which is why I am reaching out. The reason I believe the speed is slow is that I am using MACE along with other MLIPs, but MACE shows an unusually slow speed. I am attaching the log file from the current calculation as well. Although the calculation is not yet complete, it seems to be running at a very low speed.

[lammps_1000step.log](https://github.com/user-attachments/files/16694928/lammps_1000step.log)

**Reply by Juo-kim** [reply]
> Would you please try again, running 1000 steps, and send us the log? It is normal for the first several steps to be extremely slow while Torch is adjusting the sizes of its internal arrays.
> 
> > you have a huge number of ghost atoms, and that dominates the calculation
> 
> This is indeed a factor, but in `no_domain_decomposition` mode (which they are using) the situation is much improved because the ghost atoms are mapped back to local atoms.

@wcwitt 
Additionally, I will upload the results from the 1000-step calculation. Although there has been some improvement in the calculation results, I am still not entirely satisfied. Is this a reasonable level of performance?

Then, would disabling no_domain_decomposition help improve the speed? I had enabled this option because I was told that it would help improve the speed.

Similarly, I will also attach the results from the actual calculations I am using. I would appreciate it if you could take a look.

[lammps_1000step.log](https://github.com/user-attachments/files/16694968/lammps_1000step.log)

**Reply by gabor1** [reply]
no @wcwitt is right, keep the no_domain_decomposition. 

your 1000 step run now shows 0.5 ns/day, so more than 2.5x the speed you had for your 10 step run. Why are you using this particular L2 mace? with its large cutoff (6 A) and large messages, it will be significantly slower than the medium (L1) models. Can you say what sort of speed you were expecting? What other MLIPs are you comparing to? For example, if you are comparing to a message passing model with invariant messages, then the L0 mace is the equivalent of that, and it would run several times faster than this L2 mace that you are using - at the cost of some accuracy of course, but you'd need to establish what accuracy is sufficient for your application.

**Reply by gabor1** [reply]
If you are comparing to something like ACE or MTP, those could be 10-20x faster than your current run (depending on their settings), but that is just the consequence of those being linear models. MACE is nonlinear, and is inherently slower than linear models - but for this, you get much enhanced transferability and MD stability.

**Reply by gabor1** [reply]
Maybe this is also a problem: 

`Neighbor list info ...
  update: every = 1 steps, delay = 0 steps, check = yes
`

You are instructing lammps to recompute the neighbour list every step ?

**Reply by Juo-kim** [reply]
> You are instructing lammps to recompute the neighbour list every step ?

The reason I am using the L2 model is that I previously compared the speed using the 2023-12-03-mace-mp.model-lammps, and there was no speed difference between the two models. The models I am comparing are M3GNet and Sevennet, and although it might not be entirely appropriate to compare the speed with different models, the speed difference seems to be several times greater, so I wanted to check and inquire about it.

As for "neighbor list", that was not my intention. I will correct this and run the calculation again .Thank you for your help.

**Reply by gabor1** [reply]
M3GNet is an invariant message passing model I think, and Sevennet is basically NEquip. on a single GPU, it's quite straightforward to compare MACE with NEquip, because they use similar operations. What is the cutoff in your Sevennet model?

**Reply by gabor1** [reply]
the L1 should be about 2x faster than the L2 and the L0 model again about 2x faster.

**Reply by gabor1** [reply]
BTW these are the latest models (small and medium), they are improved version, including pair repulsion (so more stable)

https://github.com/ACEsuit/mace-mp/releases/tag/mace_mp_0b

**Reply by Juo-kim** [reply]
I also thought that the speed would be similar to SevenNet, but it was 5 times faster in terms of timestep/s.
The cutoff in my sevennet model was 7 and it was also pretrained model.

I wasn't aware that there was a new model, but I'll try using it.

As for the neighbor list, I didn't change anything specific. Isn't it supposed to update at every step by default?

**Reply by gabor1** [reply]
I can't guarantee that it will make a big difference but it is customary to set a "skin", so that the neighbour list doesn't get built unless atoms move sufficiently far: `neighbor skin style`, e.g. `neighbour 0.5 bin` (see here: https://docs.lammps.org/neighbor.html)

**Reply by Juo-kim** [reply]
Even if I change the neighbor setting, it seems like the neighbor list still updates at every step. 
But when I tried using the small and medium models, the speed improved.
I think the model was too heavy to what I expected.

[lammps_small.log](https://github.com/user-attachments/files/16695988/lammps_small.log)
[lammps_medium.log](https://github.com/user-attachments/files/16695989/lammps_medium.log)

**Reply by gabor1** [reply]
I think what happens (if my parsing of the LAMMPS doc is right) that it CHECKS every time whether the neighbourlist needs to be updated (based on the movement of atoms and the skin you specified) and it only rebuilds if necessary.

**Reply by gabor1** [reply]
I'm looking at the supplementary of the sevennet paper (https://pubs.acs.org/doi/10.1021/acs.jctc.4c00190) and they show that for a single GPU, sevennet is 1.5 times faster than mace (they don't say which mace, but it is likely the medium one)

**Reply by gabor1** [reply]
they also show that the multi-gpu implementation of mace is not efficient, which we know, and we are working on it.

**Reply by Juo-kim** [reply]
Yes, Considering that the speed may vary depending on the structure or environment of lammps, it seems to make sense. 
Thank you for your attention!

**Reply by ilyes319** [reply]
@Juo-kim What kind of speed do you get with seven net on a single GPU in terms of ns/day? I would be very surprised is 7net is 5x faster than the L2 model for something like 200 atoms.

**Reply by liudfgoo** [reply]
Recently, I encountered the same problem. And I think it's related to the **cutoff distance** settings in the lammps. In the file uploaded by @Juo-kim, the `neighbor/atoms` and  `distance cutoff` are below,
```
  update: every = 1 steps, delay = 0 steps, check = yes
  max neighbors/atom: 2000, page size: 100000
  master list distance cutoff = 14
  ghost atom cutoff = 14
```
Logically, the cutoff radius here should align with what was used in model training. And I want to know how to set the radius here. Looking forward to your help!
